File: implementations/dotnet/CrossVerify.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Ovidiu Ancuta
//
// aioschema/dotnet v0.5.6 | AIOSchema spec v0.5.6
// https://aioschema.org

using System.Text;
using System.Text.Json;
using AIOSchema;

namespace AIOSchema;

public static class CrossVerify
{
    public static int Run(string vectorsPath)
    {
        if (!File.Exists(vectorsPath))
        {
            Console.Error.WriteLine($"cross_verify_vectors.json not found at: {vectorsPath}");
            Console.Error.WriteLine("Set AIOSCHEMA_VECTORS env var or pass path as argument.");
            return 1;
        }

        Console.WriteLine("AIOSchema v0.5.6 — Cross-Verification Suite (.NET)");
        Console.WriteLine("===================================================");
        Console.WriteLine($"Vectors: {vectorsPath}");
        Console.WriteLine();

        using var doc     = JsonDocument.Parse(File.ReadAllText(vectorsPath));
        var specVersion   = doc.RootElement.GetProperty("spec_version").GetString();
        var vectors       = doc.RootElement.GetProperty("vectors");
        Console.WriteLine($"Spec version: {specVersion}");
        Console.WriteLine();

        int passed = 0;
        int failed = 0;

        foreach (var v in vectors.EnumerateArray())
        {
            var id     = v.GetProperty("id").GetString()!;
            var name   = v.GetProperty("name").GetString()!;
            var inputs = v.GetProperty("inputs");

            try
            {
                RunVector(id, name, inputs, v, ref passed, ref failed);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  FAIL  {id}: {name}");
                Console.WriteLine($"        Exception: {ex.Message}");
                failed++;
            }
        }

        Console.WriteLine();
        Console.WriteLine(new string('=', 51));
        Console.WriteLine($"Results: {passed} passed, {failed} failed, {passed + failed} total");
        Console.WriteLine(new string('=', 51));

        if (failed > 0)
        {
            Console.Error.WriteLine($"\nFAIL: {failed} cross-verification vector(s) failed.");
            return 1;
        }

        Console.WriteLine("\nAll cross-verification vectors passed.");
        Console.WriteLine(".NET implementation is interoperable with Python / TypeScript / Node.js / Go / Rust.");
        return 0;
    }

    private static void RunVector(
        string id, string name, JsonElement inputs, JsonElement v,
        ref int passed, ref int failed)
    {
        // CV-01 to CV-04: hash computation
        if (inputs.TryGetProperty("data_hex", out var dataHexEl) &&
            inputs.TryGetProperty("algorithm", out var algEl))
        {
            var data     = Convert.FromHexString(dataHexEl.GetString()!);
            var alg      = algEl.GetString()!;
            var expected = v.GetProperty("expected").GetString()!;
            var result   = Algorithms.ComputeHash(data, alg);

            if (result == expected)
            {
                Console.WriteLine($"  PASS  {id}: {name}");
                passed++;
            }
            else
            {
                Console.WriteLine($"  FAIL  {id}: {name}");
                Console.WriteLine($"        expected: {expected}");
                Console.WriteLine($"        got:      {result}");
                failed++;
            }
            return;
        }

        // CV-05: canonical JSON
        if (inputs.TryGetProperty("object", out var objEl))
        {
            var expected = v.GetProperty("expected").GetString()!;
            var dict     = JsonElementToDict(objEl);
            var result   = Algorithms.CanonicalJson(dict);

            if (result == expected)
            {
                Console.WriteLine($"  PASS  {id}: {name}");
                passed++;
            }
            else
            {
                Console.WriteLine($"  FAIL  {id}: {name}");
                Console.WriteLine($"        expected: {expected}");
                Console.WriteLine($"        got:      {result}");
                failed++;
            }
            return;
        }

        // CV-06: core_fingerprint
        if (inputs.TryGetProperty("core_fields", out var cfEl))
        {
            var expected = v.GetProperty("expected").GetString()!;
            var fields   = new Dictionary<string, object?>();
            foreach (var key in Constants.CoreHashFields)
                if (cfEl.TryGetProperty(key, out var val))
                    fields[key] = val.GetString();

            var canonical = Algorithms.CanonicalBytes(fields);
            var result    = Algorithms.ComputeHash(canonical, "sha256");

            if (result == expected)
            {
                Console.WriteLine($"  PASS  {id}: {name}");
                passed++;
            }
            else
            {
                Console.WriteLine($"  FAIL  {id}: {name}");
                Console.WriteLine($"        expected: {expected}");
                Console.WriteLine($"        got:      {result}");
                failed++;
            }
            return;
        }

        // CV-07 to CV-14: manifest verification
        if (inputs.TryGetProperty("asset_hex", out var assetHexEl) &&
            inputs.TryGetProperty("manifest", out var manifestEl))
        {
            var assetBytes       = Convert.FromHexString(assetHexEl.GetString()!);
            var manifest         = ParseManifest(manifestEl);
            var result           = Verifier.Verify(assetBytes, manifest);
            var expectedSuccess  = v.GetProperty("expected").GetProperty("success").GetBoolean();

            bool ok = result.Success == expectedSuccess;
            // If expected has match_type, check it too
            if (ok && v.GetProperty("expected").TryGetProperty("match_type", out var mtEl))
                ok = result.MatchType == mtEl.GetString();

            if (ok)
            {
                Console.WriteLine($"  PASS  {id}: {name}");
                if (result.Warnings.Count > 0)
                    foreach (var w in result.Warnings)
                        Console.WriteLine($"        warn: {w}");
                passed++;
            }
            else
            {
                Console.WriteLine($"  FAIL  {id}: {name}");
                Console.WriteLine($"        expected success={expectedSuccess}, got={result.Success}");
                Console.WriteLine($"        message: {result.Message}");
                failed++;
            }
            return;
        }

        // Unknown vector shape
        Console.WriteLine($"  SKIP  {id}: {name} (unrecognized input shape)");
    }

    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------

    private static AIOSManifest ParseManifest(JsonElement el)
    {
        var coreEl = el.GetProperty("core");
        var c      = new CoreBlock();

        string? GetStr(string k) =>
            coreEl.TryGetProperty(k, out var v) && v.ValueKind != JsonValueKind.Null
                ? v.GetString() : null;

        c.AssetId               = GetStr("asset_id") ?? "";
        c.SchemaVersion         = GetStr("schema_version") ?? "";
        c.CreationTimestamp     = GetStr("creation_timestamp") ?? "";
        c.CreatorId             = GetStr("creator_id") ?? "";
        c.CoreFingerprint       = GetStr("core_fingerprint");
        c.HashSchemaBlock       = GetStr("hash_schema_block");
        c.Signature             = GetStr("signature");
        c.ManifestSignature     = GetStr("manifest_signature");
        c.AnchorReference       = GetStr("anchor_reference");
        c.PreviousVersionAnchor = GetStr("previous_version_anchor");

        if (coreEl.TryGetProperty("hash_original", out var ho))
        {
            if (ho.ValueKind == JsonValueKind.Array)
                c.HashOriginal = ho.EnumerateArray().Select(e => e.GetString()!).ToList<string>();
            else
                c.HashOriginal = ho.GetString()!;
        }

        var ext = new Dictionary<string, object?>();
        if (el.TryGetProperty("extensions", out var extEl) && extEl.ValueKind == JsonValueKind.Object)
            foreach (var p in extEl.EnumerateObject())
                ext[p.Name] = JsonToObject(p.Value);

        return new AIOSManifest { Core = c, Extensions = ext };
    }

    private static Dictionary<string, object?> JsonElementToDict(JsonElement el)
    {
        var d = new Dictionary<string, object?>();
        foreach (var p in el.EnumerateObject())
            d[p.Name] = JsonToObject(p.Value);
        return d;
    }

    private static object? JsonToObject(JsonElement je) => je.ValueKind switch
    {
        JsonValueKind.Null   => null,
        JsonValueKind.True   => true,
        JsonValueKind.False  => false,
        JsonValueKind.String => je.GetString(),
        JsonValueKind.Number => je.TryGetInt64(out var l) ? (object)l : je.GetDouble(),
        JsonValueKind.Array  => je.EnumerateArray().Select(e => JsonToObject(e)).ToList<object?>(),
        JsonValueKind.Object => JsonElementToDict(je),
        _ => null,
    };
}
// -- end aioschema/dotnet v0.5.6 | AIOSchema spec v0.5.6 --

File: implementations/rust/src/cross_verify.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Ovidiu Ancuta
//
// aioschema/rust v0.5.6 | AIOSchema spec v0.5.6
// https://aioschema.org

use aioschema::{
    algorithms::canonical_core_fields,
    canonical_json, compute_hash, parse_hash_prefix, verify_manifest, Manifest,
    VerifyOptions, CORE_HASH_FIELDS,
};
use base64;
use base64::Engine;
use hex;
use serde::Deserialize;
use serde_json::Value;
use std::path::PathBuf;

// ── Vector file types ─────────────────────────────────────────────────────────

#[derive(Deserialize)]
struct VectorFile {
    spec_version: String,
    vectors:      Vec<Vector>,
}

#[derive(Deserialize)]
struct Vector {
    id:          String,
    name:        String,
    inputs:      Value,
    expected:    Value,
}

// ── Loader ────────────────────────────────────────────────────────────────────

fn load_vectors() -> VectorFile {
    // Prefer explicit env var
    if let Ok(p) = std::env::var("AIOSCHEMA_VECTORS") {
        let data = std::fs::read_to_string(&p)
            .unwrap_or_else(|e| panic!("cannot read AIOSCHEMA_VECTORS={p}: {e}"));
        return serde_json::from_str(&data).expect("parse vectors JSON");
    }

    // Probe relative paths from CARGO_MANIFEST_DIR
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let candidates = [
        manifest_dir.join("../../cross_verify_vectors.json"),
        manifest_dir.join("../cross_verify_vectors.json"),
        manifest_dir.join("cross_verify_vectors.json"),
        PathBuf::from("cross_verify_vectors.json"),
    ];
    for path in &candidates {
        if path.exists() {
            let data = std::fs::read_to_string(path)
                .unwrap_or_else(|e| panic!("read {}: {e}", path.display()));
            return serde_json::from_str(&data).expect("parse vectors JSON");
        }
    }
    panic!(
        "cross_verify_vectors.json not found; \
         set AIOSCHEMA_VECTORS env var to its path"
    );
}

// ── Main test ─────────────────────────────────────────────────────────────────

#[test]
fn test_cross_verify_all_vectors() {
    let vf = load_vectors();
    eprintln!(
        "Loaded {} vectors (spec_version={})",
        vf.vectors.len(),
        vf.spec_version
    );
    assert!(vf.vectors.len() >= 14, "expected at least 14 CV vectors");

    let mut failures = 0usize;

    for v in &vf.vectors {
        let ok = match v.id.as_str() {
            "CV-01" | "CV-02" | "CV-03" | "CV-04" => run_hash_vector(v),
            "CV-05"                                => run_canonical_json_vector(v),
            "CV-06"                                => run_core_fingerprint_vector(v),
            id if id.starts_with("CV-")            => run_manifest_verify_vector(v),
            _                                      => {
                eprintln!("UNKNOWN vector id {}", v.id);
                false
            }
        };
        if ok {
            eprintln!("  PASS  {} — {}", v.id, v.name);
        } else {
            eprintln!("  FAIL  {} — {}", v.id, v.name);
            failures += 1;
        }
    }

    assert_eq!(failures, 0, "{failures} vector(s) failed");
}

// ── CV-01..CV-04: Hash computation ───────────────────────────────────────────

fn run_hash_vector(v: &Vector) -> bool {
    let data_hex = v.inputs["data_hex"].as_str().unwrap_or("");
    let algorithm = v.inputs["algorithm"].as_str().unwrap();
    let expected  = v.expected.as_str().unwrap();

    let data = hex::decode(data_hex).expect("decode data_hex");
    match compute_hash(&data, algorithm) {
        Ok(got) => {
            if got != expected {
                eprintln!("    got  {got}");
                eprintln!("    want {expected}");
                false
            } else {
                true
            }
        }
        Err(e) => {
            eprintln!("    ComputeHash error: {e}");
            false
        }
    }
}

// ── CV-05: Canonical JSON ─────────────────────────────────────────────────────

fn run_canonical_json_vector(v: &Vector) -> bool {
    let obj      = &v.inputs["object"];
    let expected = v.expected.as_str().unwrap();

    match canonical_json(obj) {
        Ok(bytes) => {
            let got = String::from_utf8(bytes).unwrap();
            if got != expected {
                eprintln!("    got  {got}");
                eprintln!("    want {expected}");
                false
            } else {
                true
            }
        }
        Err(e) => {
            eprintln!("    CanonicalJSON error: {e}");
            false
        }
    }
}

// ── CV-06: core_fingerprint ───────────────────────────────────────────────────

fn run_core_fingerprint_vector(v: &Vector) -> bool {
    let core_fields = &v.inputs["core_fields"];
    let expected    = v.expected.as_str().unwrap();

    // Determine algorithm from expected value
    let (alg, _) = match parse_hash_prefix(expected) {
        Ok(pair) => pair,
        Err(e) => {
            eprintln!("    parse expected hash: {e}");
            return false;
        }
    };

    // Build a Value::Object containing only CORE_HASH_FIELDS
    let mut subset = serde_json::Map::new();
    if let Some(obj) = core_fields.as_object() {
        for &field in CORE_HASH_FIELDS {
            if let Some(val) = obj.get(field) {
                subset.insert(field.to_string(), val.clone());
            }
        }
    }
    let subset_val = Value::Object(subset);

    let canonical = match canonical_core_fields(&subset_val) {
        Ok(b) => b,
        Err(e) => {
            eprintln!("    canonical_core_fields error: {e}");
            return false;
        }
    };

    match compute_hash(&canonical, alg) {
        Ok(got) => {
            if got != expected {
                eprintln!("    got  {got}");
                eprintln!("    want {expected}");
                false
            } else {
                true
            }
        }
        Err(e) => {
            eprintln!("    ComputeHash error: {e}");
            false
        }
    }
}

// ── CV-07..CV-18: Manifest verification ────────────────────────────────────

fn run_manifest_verify_vector(v: &Vector) -> bool {
    let asset_hex     = v.inputs["asset_hex"].as_str().unwrap();
    let manifest_val  = &v.inputs["manifest"];
    let exp_success   = v.expected["success"].as_bool().unwrap();

    let asset_data = match hex::decode(asset_hex) {
        Ok(b) => b,
        Err(e) => {
            eprintln!("    decode asset_hex: {e}");
            return false;
        }
    };

    let manifest: Manifest = match serde_json::from_value(manifest_val.clone()) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("    deserialize manifest: {e}");
            if !exp_success {
                return true;
            }
            return false;
        }
    };

    // Decode public key if provided
    let pub_key_bytes: Option<Vec<u8>> = v.inputs
        .get("public_key_b64")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| base64::engine::general_purpose::STANDARD.decode(s).unwrap());
    let opts = VerifyOptions {
        public_key: pub_key_bytes.as_deref(),
        ..VerifyOptions::default()
    };

    let result = match verify_manifest(&asset_data, &manifest, &opts) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("    verify_manifest returned Err: {e}");
            return false;
        }
    };

    if result.success != exp_success {
        eprintln!(
            "    success={} (want {}): {}",
            result.success, exp_success, result.message
        );
        return false;
    }

    // If vector specifies match_type, check it when successful
    if let Some(exp_mt) = v.expected.get("match_type").and_then(|v| v.as_str()) {
        if result.success {
            let got_mt = result
                .match_type
                .as_ref()
                .map(|m| m.to_string())
                .unwrap_or_default();
            if got_mt != exp_mt {
                eprintln!("    match_type={got_mt} (want {exp_mt})");
                return false;
            }
        }
    }

    // If vector specifies message_contains, check it
    if let Some(exp_msg) = v.expected.get("message_contains").and_then(|v| v.as_str()) {
        if !result.message.contains(exp_msg) {
            eprintln!("    message does not contain {exp_msg}: {}", result.message);
            return false;
        }
    }

    true
}
// -- end aioschema/rust v0.5.6 | AIOSchema spec v0.5.6 --
